### PR TITLE
Additional set of Thread Sanitizer related fixes

### PR DESCRIPTION
**Pull request by alexey-katranov**
* Fix incorrect assert in `concurrent_hash_map` implementation
* Fix incorrect asserts in `global_control` implementation
* Fix exception pointer inside `task_group_context`
* Rework racy asserts in test_flow_graph_whitebox 
* Improve `test_eh_algorithm` to work with sanitizers 
* Disable memory limit test in `test_malloc_compliance`
* Add lifetime asserts in `collaborative_call_once`
* Rework `backref` space in `tbbmalloc` to use atomics
* Fix a lifetime issue in `utils::SpinBariier`
* Some tests were disabled

### CODE REVIEW

File: include/oneapi/tbb/concurrent_hash_map.h [hunk 1]
```diff
@@ -724,7 +724,7 @@ class concurrent_hash_map
 
     // TODO refactor to hash_base
     void rehash_bucket( bucket *b_new, const hashcode_type hash ) {
-        __TBB_ASSERT( *(intptr_t*)(&b_new->mutex), "b_new must be locked (for write)");
+        __TBB_ASSERT( !b_new->mutex.try_lock(), "b_new must be locked (for write)");
```
Comment (alexey-katranov) [inline include/oneapi/tbb/concurrent_hash_map.h:727]:
```suggestion
```

File: test/tbb/test_flow_graph_whitebox.cpp [hunk 2]
```diff
@@ -277,7 +277,10 @@ void TestFunctionNode() {
     });
     utils::SpinWaitWhileEq(serial_fn_state0, 0); // waiting rejecting node to start
     utils::SpinWaitWhileEq(rejected, false);
-    CHECK(fnode0.my_predecessors.empty() == false);
+    // TODO: the assest below is not stable due to the logical race between try_put(1)
+    // try_put(2) and wait_for_all.
+    // Additionaly, empty() cannot be called concurrently due to null_mutex used in implementaton/
```
Comment (alexey-katranov) [inline test/tbb/test_flow_graph_whitebox.cpp:282]:
```suggestion
    // Additionally, empty() cannot be called concurrently due to null_mutex used in implementation
```

File: test/tbb/test_flow_graph_whitebox.cpp [hunk 3]
```diff
@@ -323,12 +326,14 @@ void TestFunctionNode() {
     utils::SpinWaitWhileEq(serial_fn_state0, 0); // waiting rejecting node to start
     // wait for the edges to reverse
     utils::SpinWaitWhileEq(rejected, false);
-    CHECK(fnode0.my_predecessors.empty() == false);
+    // TODO: the assest below is not stable due to the logical race between try_put(1)
+    // try_put(2) and wait_for_all.
+    // Additionaly, empty() cannot be called concurrently due to null_mutex used in implementaton/
```
Comment (alexey-katranov) [inline test/tbb/test_flow_graph_whitebox.cpp:331]:
```suggestion
    // Additionally, empty() cannot be called concurrently due to null_mutex used in implementation
```

**Review by vlserov — DISMISSED** [review]
LGTM

File: include/oneapi/tbb/collaborative_call_once.h [hunk 4]
```diff
@@ -185,6 +185,8 @@ class collaborative_once_flag : no_copy {
                     // The moonlighting threads are not expected to handle exceptions from user functor.
                     // Therefore, no exception is expected from assist().
                     shared_runner->assist();
+                    // Read m_state while holding the guard because m_state can be destroyed after releasing the guard
+                    expected = m_state.load(std::memory_order_relaxed);
```
Comment (alexey-katranov) [inline include/oneapi/tbb/collaborative_call_once.h:189]: @BenFrantzDale, pay attention to the fix if you already started using the functionality (in worst case, the implementation without the fix might lead to memory corruptions or deadlocks).

File: test/common/spin_barrier.h [hunk 5]
```diff
@@ -128,21 +134,22 @@ class SpinBarrier {
 
         size_type epoch = myEpoch;
         size_type numThreads = myNumThreads; // read it before the increment
```
Comment (ivankochin) [inline test/common/spin_barrier.h:136]: It seems like the `numThreads` variable can be replaced by `myNumThreads`

File: test/common/spin_barrier.h [hunk 6]
```diff
@@ -128,21 +134,22 @@ class SpinBarrier {
 
         size_type epoch = myEpoch;
```
Comment (ivankochin) [inline test/common/spin_barrier.h:135]:
```suggestion
        size_type epoch = myEpoch.load(std::memory_order_relaxed);
```

File: test/common/spin_barrier.h [hunk 7]
```diff
@@ -128,21 +134,22 @@ class SpinBarrier {
...
         size_type numThreads = myNumThreads; // read it before the increment
-        int threadsLeft = static_cast<int>(numThreads - myNumThreadsFinished++ - 1);
+        int threadsLeft = static_cast<int>(numThreads - myNumThreadsFinished.fetch_add(1, std::memory_order_release) - 1);
```
Comment (ivankochin) [inline test/common/spin_barrier.h:137]:
```suggestion
        int threadsLeft = static_cast<int>(myNumThreads - myNumThreadsFinished.fetch_add(1, std::memory_order_release) - 1);
```

File: test/tbbmalloc/test_malloc_compliance.cpp [hunk 8]
```diff
@@ -999,22 +993,27 @@ void CMemTest::TestAlignedParameters()
     }
 }
 
-void CMemTest::RunAllTests(int total_threads)
+void CMemTest::RunAllTests(int )
```
Comment (ivankochin) [inline test/tbbmalloc/test_malloc_compliance.cpp:996]: ?

**Review by pavelkumbrasev — COMMENTED** [review]
Apply remark comments

File: test/tbb/test_eh_thread.cpp [hunk 9]
```diff
@@ -33,9 +33,24 @@
...
+#include <sys/resource.h>
+
+void limitThreads(size_t limit)
+{
+    rlimit rlim;
+
+    int ret = getrlimit(RLIMIT_NPROC, &rlim);
+    CHECK_MESSAGE(0 == ret, "getrlimit has returned an error");
+
+    rlim.rlim_cur = rlim.rlim_max == (rlim_t)RLIM_INFINITY ? limit : utils::min(limit, rlim.rlim_max);
```
Comment (ivankochin) [inline test/tbb/test_eh_thread.cpp:49]: Just a minor comment to improve readability:
```suggestion
    rlim.rlim_cur = (rlim.rlim_max == (rlim_t)RLIM_INFINITY) ? limit : utils::min(limit, rlim.rlim_max);
```

**Review by ivankochin — DISMISSED** [review]

File: test/tbbmalloc/test_malloc_compliance.cpp (hunk 8 above)
Comment (alexey-katranov) [reply test/tbbmalloc/test_malloc_compliance.cpp]: Removed debug stuff

**Review by pavelkumbrasev — APPROVED** [review]

**Pull request merged**